sep: change disbursement to be safer with multiple and regulated assets (#78)
stellar/experimental-payment-channels: specifications/sep-payment-channel-transactions.md

all previous declaration, payment, and closing transaction and that will
disburse $8 to I and $22 to R.

To update the payment channel state, the participants:

1. Increment i.
2. Sign and exchange a closing transaction C_i.
3. Sign and exchange zero or more payment transactions P_{i,a}.
4. Sign and exchange a declaration transaction D_i.

It is important that D_i is signed after C_i and P_i because D_i will invalidate
any previously signed C_i and P_i. If I and R signed and exchanged D_i first
either party could prevent the channel from closing without coordination by
submitting D_i and refusing to sign C_i.  The participants would not be able to
close the channel, or regain control of the accounts, and the assets within
without coordinating with each other.

The transactions are constructed as follows:

- C_i, the _closing transaction_, changes the signing weights on EI such that I
unilaterally controls EI, and the signing weights on ER such that R unilaterally
controls ER.  C_i has source account EI, sequence number s_i+1+n.

  The `minSeqAge` and `minSeqLedgerGap` prevents a misbehaving party from
  executing C_i when the channel state has already progressed to a later

>   iteration number, as the other party has the period 2O to invalidate C_i by

Reviewer: 2O is hard to distinguish from 20 (twenty) when reading:
```suggestion
  iteration number, as the other party has the period 2*O to invalidate C_i by
```

Author: Oops, 2O should be gone, it should be O. 2O was required in an earlier commit on this PR, then I removed it. Fix incoming.

Author: I've fixed this in 9b78df3.

Although I realize this creates a new issue, if there are no payments, the close is not delayed. I'll fix that shortly.

Author: I fixed the second issue with the observation period missing on close with no payments in 9fa8ec1.

Reviewer: cool, thanks for changing

  submitting D_i' for some i' > i.
  
  C_i contains operations:
  - One or more `SET_OPTIONS` operation adjusting escrow account EI's thresholds
  to give I full control of EI, and removing R's signers.
  - One or more `SET_OPTIONS` operation adjusting reserve account ER's
  thresholds to give R full control of ER, and removing I's signers.

- P_{i,a}, the _payment transaction_, disburses a single asset a from EI to ER
or from ER to EI. P_{i,a} has source account EI, sequence number s_i+1+a, where
a is the zero index of the asset where each asset in the channel is assigned an

> index. P_{i,0}, the first payment, is assigned a `minSeqAge` of O (the

Reviewer: "O" defining both `observation period time duration` and `the observation period ledger count` is confusing to me since it's implying that "O" would be the same value for both (which I don't think is the case). It looks like we're using "O" generally in this SEP to describe "observation period", which is describing two fields.

💭  proposal to change "O" here to `O_minSeqAge` and `O_minSeqLedgerGap`, or some other way of differentiating between its 2 uses

Author: Agree it is a little confusing, although I'm not sure there's a good formal notation like what you propose. The correct way to represent the two different values would be to define the separately, such as to define them as OA and OG. I'm not sure it is worth introducing two separate variables though. It's out of scope of this PR. Could you open an issue?

Reviewer: https://github.com/stellar/experimental-payment-channels/issues/79

For example, if the channel initial state is $30 of which $10 belongs to I and
$20 belongs to R, the first closing transaction will disburse $10 to I and $20
to R. If I makes a payment of $2 to R, this update process will involve I and R
agreeing on new declaration, payment, and closing transactions that supersede
all previous declaration, payment, and closing transaction and that will
disburse $8 to I and $22 to R.

To update the payment channel state, the participants:

1. Increment i.
2. Sign and exchange a closing transaction C_i.
3. Sign and exchange zero or more payment transactions P_{i,a}.
4. Sign and exchange a declaration transaction D_i.

> It is important that D_i is signed after C_i and P_i because D_i will invalidate

Author: I think this is a good change, but I think we should add some specificity to what's being exchanged. They shouldn't exchange signatures, but they may still exchange transactions.

```suggestion
It is critical that signatures for D_i are exchanged after C_i and P_i because D_i will invalidate
```

Wdyt?

> It is important that D_i is signed after C_i and P_i because D_i will invalidate

Author: I took this change to a new PR since it is independent to the change this PR is introducing. Could you review https://github.com/stellar/experimental-payment-channels/pull/123?

> It is important that D_i is signed after C_i and P_i because D_i will invalidate

Reviewer 2: Sure, thanks!
